lib/versioning/VersionID.js

```diff
@@ -0,0 +1,136 @@
+ *
+ * @param {Number} span - time to wait in nanoseconds (1/1000000 millisecond)
+ * @return {Undefined} - nothing
+ */
+function wait(span) {
+    function getspan(diff) {
+        return diff[0] * 1e9 + diff[1];
+    }
+    const start = process.hrtime();
+    while (getspan(process.hrtime(start)) < span);
+}
+
+/**
+ * This function returns a "versionId" string indicating the current time as a
+ * combination of the current time in millisecond, the position of the request
+ * in that millisecond, and the identifier of the local site (which could be
+ * datacenter, region, or server depending on the notion of geographics). This
+ * function is stateful which means it keeps some values in the memory and the
+ * next call depends on the previous call.
+ *
+ * @param {string} info - the additional info to ensure uniqueness if desired
+ * @return {string} - the formated versionId string
+ */
+function generateVersionId(info) {
+    // Need to wait for the millisecond slot got "flushed". We wait for
+    // only a single millisecond when the module is restarted, which is
+    // necessary for the correctness of the system. This is therefore cheap.
+    if (prvts === 0) {
+        wait(1000000);
+    }
+    // get the present epoch (in millisecond)
+    const ts = Date.now();
+    // A bit more rationale: why do we use a sequence number instead of using
+    // process.hrtime which gives us time in nanoseconds? The idea is that at
+    // any time resolution, some concurrent requests may have the same time due
+    // to the way the OS is queueing requests or getting clock cycles. Our
+    // approach however will give the time based on the position of a request
+    // in the queue for the same millisecond which is supposed to be unique.
+
+    // increase the position if this request is in the same epoch
+    prvsq = (prvts === ts) ? prvsq + 1 : 0;
+    prvts = ts;
+
+    // In the default cases, we reverse the chronological order of the
+    // timestamps so that all versions of an object can be retrieved in the
+    // reversed chronological order---newest versions first. This is because of
+    // the limitation of leveldb for listing keys in the reverse order.
+    return `${TEMPLATE_TS}${MAX_TS - prvts}`.slice(-LENGTH_TS) +
+        `${TEMPLATE_SQ}${MAX_SQ - prvsq}`.slice(-LENGTH_SQ) + SITE_ID + info;
+}
+
+/**
+ * Encrypt a versionId to add a validation layer before processing
+ * bucket versioning related requests.
+ *
+ * @param {string} str - the versionId to encrypt
+ * @return {string} - the encrypted versionId
+ */
+function encrypt(str) {
+    const cipher = crypto.createCipher(
+            VID_CRYPTO_ALGORITHM, VID_CRYPTO_PASSWORD);
+    let encrypted = cipher.update(str,
+            VID_CRYPTO_ENCODING_TXT, VID_CRYPTO_ENCODING);
+    encrypted += cipher.final(VID_CRYPTO_ENCODING);
+    let length = encrypted.length;
+    while (length && encrypted.charAt(length - 1) === '=') {
```

Comment:
Yes it was not carefully crafted. Some other alternatives can be used like a simple `indexOf`, but regex seems concise and is a good solution.